egs/librispeech/asr/simple_v1/mmi_mbr_train.py

```diff
@@ -0,0 +1,591 @@
+    finite_indexes = torch.nonzero(mask).squeeze(1)
+    if False:
+        bad_indexes = torch.nonzero(~mask).squeeze(1)
+        if bad_indexes.shape[0] > 0:
+            print("Bad indexes: ", bad_indexes, ", bad lengths: ",
+                  frames_per_seq[bad_indexes], " vs. max length ",
+                  torch.max(frames_per_seq), ", avg ",
+                  (torch.sum(frames_per_seq) / frames_per_seq.numel()))
+    # print("finite_indexes = ", finite_indexes, ", tot_scores = ", tot_scores)
+    ok_frames = frames_per_seq[finite_indexes].sum()
+    all_frames = frames_per_seq.sum()
+    return (tot_scores[finite_indexes].sum(), ok_frames, all_frames)
+
+
+def get_loss(batch: Dict,
+             model: AcousticModel,
+             P: k2.Fsa,
+             device: torch.device,
+             graph_compiler: MmiMbrTrainingGraphCompiler,
+             is_training: bool,
+             optimizer: Optional[torch.optim.Optimizer] = None):
+    feature = batch['features']
+    supervisions = batch['supervisions']
+    supervision_segments = torch.stack(
+        (supervisions['sequence_idx'],
+         torch.floor_divide(supervisions['start_frame'],
+                            model.subsampling_factor),
+         torch.floor_divide(supervisions['num_frames'],
+                            model.subsampling_factor)), 1).to(torch.int32)
+    indices = torch.argsort(supervision_segments[:, 2], descending=True)
+    supervision_segments = supervision_segments[indices]
+
+    texts = supervisions['text']
+    texts = [texts[idx] for idx in indices]
+    assert feature.ndim == 3
+    # print(supervision_segments[:, 1] + supervision_segments[:, 2])
+
+    feature = feature.to(device)
+    # at entry, feature is [N, T, C]
+    feature = feature.permute(0, 2, 1)  # now feature is [N, C, T]
+    if is_training:
+        nnet_output = model(feature)
+    else:
+        with torch.no_grad():
+            nnet_output = model(feature)
+
+    # nnet_output is [N, C, T]
+    nnet_output = nnet_output.permute(0, 2, 1)  # now nnet_output is [N, T, C]
+
+    if is_training:
+        num_graph, den_graph, decoding_graph = graph_compiler.compile(texts, P)
+    else:
+        with torch.no_grad():
+            num_graph, den_graph, decoding_graph = graph_compiler.compile(
+                texts, P)
+
+    assert num_graph.requires_grad == is_training
+    assert den_graph.requires_grad is False
+    assert decoding_graph.requires_grad is False
+
+    num_graph = num_graph.to(device)
+    den_graph = den_graph.to(device)
+
+    decoding_graph = decoding_graph.to(device)
+
+    dense_fsa_vec = k2.DenseFsaVec(nnet_output, supervision_segments)
+    assert nnet_output.device == device
+
+    num_lats = k2.intersect_dense(num_graph,
+                                  dense_fsa_vec,
+                                  10.0,
+                                  seqframe_idx_name='seqframe_idx')
+
+    den_lats = k2.intersect_dense(den_graph, dense_fsa_vec, 10.0)
```

Comment:
As an alternative, we can try using decoding_graph here intead of den_graph (and intersect_dense_pruned).

Reply:
So in this case, are there only two graphs: `num_graph` and `decoding_graph` ?

Reply:
Yes.  den_graph and decoding_graph are just different ways of making a generic graph.  It's empirical which one works better.

Reply:
Some sequences may return `-inf` for total scores.

Should we discard/ignore such kind of sequences?

Reply:
Yes

Reply:
Sequences with tot scores being `-inf` are discarded for MMI, should they be kept for MBR training?

Reply:
Otherwise, it is nontrivial to fix the arc indexes while creating the sparse matrices when some FSAs are removed.

Reply:
As I described it and set it up, the MBR training will automatically ignore sequences with score of -inf.

Reply:
But either way.

Reply:
wait.. sorry I was talking about the n-best stuff.  Not sure, do it the way you think is best.